cmd/node/main.go

```diff
@@ -172,152 +229,28 @@ func newConfig() (*config.Config, config.Error) {
 	viper.AutomaticEnv()
 	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
 
-	// set FlagVars first
-	if err = viper.BindPFlag("dataDir", flag.Lookup("dataDir")); err != nil {
-		log.Fatalf("failed to bind dataDir flag to viper: %v", err)
-	}
-	conf.DataDir = viper.GetString("dataDir")
-
-	if err = viper.BindPFlag("chain", flag.Lookup("chain")); err != nil {
-		log.Fatalf("failed to bind chain flag to viper: %v", err)
-	}
-	conf.Vochain.Network = viper.GetString("chain")
-
-	if err = viper.BindPFlag("dev", flag.Lookup("dev")); err != nil {
-		log.Fatalf("failed to bind dev flag to viper: %v", err)
-	}
-	conf.Dev = viper.GetBool("dev")
+	viper.BindFlagValues(pflagValueSet{flag.CommandLine})
 
-	if err = viper.BindPFlag("pprofPort", flag.Lookup("pprof")); err != nil {
-		log.Fatalf("failed to bind pprof flag to viper: %v", err)
-	}
-
-	if err = viper.BindPFlag("dbType", flag.Lookup("dbType")); err != nil {
-		log.Fatalf("failed to bind dbType flag to viper: %v", err)
-	}
-	conf.Vochain.DBType = viper.GetString("dbType")
+	conf.Vochain.DBType = viper.GetString("dbType")                 // TODO: unused?
+	conf.Vochain.Indexer.ArchiveURL = viper.GetString("archiveURL") // TODO: unused?
 
 	// use different datadirs for different networks
+	conf.DataDir = viper.GetString("dataDir")
+	conf.Vochain.Network = viper.GetString("chain")
 	conf.DataDir = filepath.Join(conf.DataDir, conf.Vochain.Network)
+	viper.Set("dataDir", conf.DataDir)
 
-	if err = viper.BindPFlag("archiveURL", flag.Lookup("archiveURL")); err != nil {
-		log.Fatalf("failed to bind archiveURL flag to viper: %v", err)
-	}
-	conf.Vochain.Indexer.ArchiveURL = viper.GetString("archiveURL")
-
-	// add viper config path (now we know it)
-	viper.AddConfigPath(conf.DataDir)
-
-	// binding flags to viper
-	if err = viper.BindPFlag("mode", flag.Lookup("mode")); err != nil {
-		log.Fatalf("failed to bind mode flag to viper: %v", err)
-	}
-	if err = viper.BindPFlag("logLevel", flag.Lookup("logLevel")); err != nil {
-		log.Fatalf("failed to bind logLevel flag to viper: %v", err)
-	}
-	if err = viper.BindPFlag("logErrorFile", flag.Lookup("logErrorFile")); err != nil {
-		log.Fatalf("failed to bind logErrorFile flag to viper: %v", err)
-	}
-	if err = viper.BindPFlag("logOutput", flag.Lookup("logOutput")); err != nil {
-		log.Fatalf("failed to bind logOutput flag to viper: %v", err)
-	}
-	if err = viper.BindPFlag("saveConfig", flag.Lookup("saveConfig")); err != nil {
-		log.Fatalf("failed to bind saveConfig flag to viper: %v", err)
-	}
-	if err = viper.BindPFlag("signingKey", flag.Lookup("signingKey")); err != nil {
```

Comment:
So this actually removes the need to bind a flag with a viper config option (that we use mainly for ENV vars and yaml config file)? I don't really understand how it works but it that's the case is a very nice improvement.

Reply:
We still bind, but now effectively in a loop rather than one by one. See these bits of code:

https://github.com/vocdoni/vocdoni-node/pull/1225/files#diff-41988d47676978b9c3637b55fba6588aed77dab7cff1aadfb41b119964785392R232

https://github.com/vocdoni/vocdoni-node/pull/1225/files#diff-41988d47676978b9c3637b55fba6588aed77dab7cff1aadfb41b119964785392R74-R105

The old code bound e.g. the `vochainPeers` flag as `vochain.Peers` on viper for the config. We now do this via the CutPrefix string handling.